docs/syntax-attributes-functions-parameters.md

```diff
@@ -0,0 +1,34 @@
+# Function Attribute Parameters
+
+## Summary
+
+This RFC proposes a syntax for function attribute parameters. This is a follow up to the [Attributes](./syntax-attributes-functions.md) RFC, which proposed a design for function attributes. We only propose a syntactic extension in this RFC; the rest of the implementation details remain unchanged.
+
+## Motivation
+
+The [Attributes](./syntax-attributes-functions.md) RFC provides a syntax for parameterless function attributes. This suffices for some potential use cases, such as `@native`, and `@inline`. However, other potential use cases such as `@deprecated` will benefit from an expressive attribute syntax that allows parameters to be supplied. We can use the parameters to provide the name of the function to be used in place of the deprecated function, and the reason for deprecation. This information can be used to generate informative deprecation warning messages. Another potential use case could be an `@unroll` attribute for loop unrolling which could use a numeric parameter for specifying the number of iterations to be unrolled. This would also require supporting attributes on loops, which would be a topic of discussion for a separate RFC. It might be desirable to allow these attributes to assume a default value for parameters not provided explicitly, but that's a discussion reserved for attribute-specific RFCs. 
+
+## Design
+
+The following syntax is proposed for attributes with parameters:
+
+```ebnf
+attribute = '@' NAME 
+          | '@' NAME '(' ATOMIC-LITERAL (',' ATOMIC-LITERAL)* ')'
+```
+
+The only extension proposed to the [Attributes](./syntax-attributes-functions.md) RFC is in the second line which allows one or more comma-separated literals to be supplied to the attribute. The `ATOMIC-LITERAL` category includes values of type `boolean`, `number`, `string`, and the `nil` value. The `nil` value can be be used for "optional" parameters when no valid value exists. As a hypothetical example, consider the `@deprecated(<new-function-name>, <warning-message>)` attribute. Assume a function is being marked deprecated because we are getting rid of the feature entirely and there is no other function replacing it. In that case, we can pass `nil` for the `<new-function-name>` parameter. An alternative would be to allow this attribute to have either one or two parameters. But that still won't help when an attribute can take three parameters and we want to ignore the second.
+
+There are three notable exclusions in this syntax. 
+
+1. Literal `table` values are excluded from being passed as parameters since they are composite structures; information supplied via table fields can be passed as separate parameters to the attribute. If they are still deemed useful, a follow-up RFC can introduce them with motivating use cases.
```

Comment:
A motivating example is literally in the paragraph above, `@deprecated`. Sometimes you want to say:

* this symbol is deprecated: `@deprecated`
* this symbol is deprecated with a replacement: `@deprecated { use = "foo" }`
* this symbol is deprecated for a reason: `@deprecated { reason = "..." }`
* this symbol is deprecated with a replacement and a reason: `@deprecated { use = "foo", reason = "..." }`

I think in the case of `@deprecated`, you really don't know which argument `use` or `reason` will be most used, so if you get the parameter ordering wrong, you end up with a lot of redundant `nil` argument in all use sites, and potentially shoot yourself in the foot because you forgot to add `nil` arguments i.e. `@deprecated("my cool reason")`.

Reply:
I chose to keep the argument type simple because attributes are not user-extensible and we have only one use case for optional attribute arguments. I deliberately decided to be conservative and leave the door open for a future RFC to introduce table syntax if really needed. In this particular instance, at max, 2 `nil` arguments need to be supplied which feels fine to me. The table syntax is bulkier than I would like.

Reply:
If we want table parameters, then there is no point in having multi-parameter attributes. Table can be used to supply all the parameters (with or without names). Attributes can be implemented to optionally accept a single table argument (without surrounding `(` and `)`).